src/shader.cpp

```diff
@@ -105,26 +106,70 @@ void Context::glLinkProgram(GLuint program)
 
             auto loc = gl.GetUniformLocation(program, uniform_name.get());
 
-            programr.uniforms.emplace(loc, uniform_info);
+            programInfo.uniforms.emplace(loc, uniform_info);
         }
     }
 }
 
 void Context::glUseProgram(unsigned int handle) {
-    auto it = programs.find(handle);
+    if (handle == 0) {
+        bound_program = 0;
+        return;
+    }
+
+    if (!validate_program_status(handle)) {
+        return;
+    }
+
+    // TODO: add optional performance warning for rebinding the same program
+    //if (handle == bound_program) {
+    //    output_fmt("glUseProgram(program = %u): Program is already bound.", handle);
+    //}
+
+    bound_program = handle;
+}
+
+void Context::glDeleteProgram(GLuint program)
+{
+    auto it = programs.find(program);
     if (it == programs.end()) {
-        output_fmt("glUseProgram(program = %u): Invalid program handle.", handle);
+        output_fmt("glUseProgram(program = %u): Invalid program handle.", program);
         return;
     }
 
-    if (it->second.link_status == LinkStatus::UNCHECKED) {
-        output_fmt("glUseProgram(program = %u): Always check program link status before trying to use the object.", handle);
+    // note: this does not change which program is bound, even if it becomes invalid here
+
+    programs.erase(it);
+}
+
+void Context::validate_program_bound(std::string_view func_name)
+{
+    if (bound_program == 0) {
+        output_fmt("%s: No program bound.", func_name.data());
```

Comment:
`output_fmt` is also not a great name. to me it just says "format some output" or "output formatted"... why not `printMessage`, messages are per se formatted

Reply:
I agree, but I think the changes to output_fmt should go in another PR.